framework/src/meshgenerators/XYZDelaunayGenerator.C

```diff
@@ -340,36 +420,95 @@ XYZDelaunayGenerator::generate()
+    UnstructuredMesh & hole_mesh = dynamic_cast<UnstructuredMesh &>(**_hole_ptrs[hole_i]);
+    auto & hole_boundary_info = hole_mesh.get_boundary_info();
+
+    // Our algorithm here requires a serialized Mesh.  To avoid
+    // redundant serialization and deserialization (libMesh
+    // MeshedHole and stitch_meshes still also require
+    // serialization) we'll do the serialization up front.
+    libMesh::MeshSerializer serial_hole(hole_mesh);
 
-            // I'd love to assert that we don't have any missing
-            // matches, but our holes might themselves have holes
-            if (it != mesh_faces.end())
+    // We'll look for any sides that match between the hole mesh and
+    // the newly triangulated mesh, and apply bcids accordingly.
+    for (auto elem : hole_mesh.element_ptr_range())
+      for (auto s : make_range(elem->n_sides()))
+        if (!elem->neighbor_ptr(s))
+        {
+          auto points = sorted_point_tuple(*elem, s);
+          auto it = mesh_faces.find(points);
+
+          // I'd love to assert that we don't have any missing
+          // matches, but our holes might themselves have holes
+          if (it != mesh_faces.end())
+          {
+            auto [main_elem, main_side] = it->second.first;
+            if (hole_i < _stitch_holes.size() && _stitch_holes[hole_i])
             {
-              auto [main_elem, main_side] = it->second;
               hole_boundary_info.add_side(elem, s, new_hole_bcid);
               mesh_boundary_info.add_side(main_elem, main_side, inner_bcid);
             }
+            // We would like to take this opportunity to identify the hole boundary
+            // We set the boundary id to hole_i + 1 at this stage (the default)
+            mesh_boundary_info.add_side(main_elem, main_side, hole_i + 1);
+            it->second.second = true;
           }
+        }
+  }
 
+  // Any sides that do not match the hole surfaces belong to the external boundary
+  // We set the boundary id to 0 at this stage (the default)
+  for (auto & [points, elem_side] : mesh_faces)
+    if (!elem_side.second)
+    {
+      auto [main_elem, main_side] = elem_side.first;
+      mesh_boundary_info.add_side(main_elem, main_side, 0);
+    }
+
+  // Now hole boundary ids are 1,2,..., N
+  // Now external boundary id is 0
+  // We will find the upper bound of the boundary ids of the mesh so that we would not overwrite
+  // things during renumbering (because of the bcids used for stitching, this is different from
+  // end_bcid)
+  const auto temp_bcid_shift = MooseMeshUtils::getNextFreeBoundaryID(*mesh);
+  // If we stitch holes, we want to make sure the netgen mesh has no conflicting boundary ids
+  // with the hole meshes.
+  // If we assign custom boundary ids, we want to make sure no overwriting occurs.
+  // So we shift all these boundary ids by temp_bcid_shift + free_boundary_id
+  // before we assign the new boundary ids
+  if (_stitch_holes.size() || output_boundary_id.size())
```

Comment:
```suggestion
  if (_stitch_holes.size() && output_boundary_id.size())
```
? if we dont stitch there is no problem

Reply:
This should be "OR":
if we need to stitch the hole meshes, we need to change the default ids to avert conflicts with ids of the hole meshes;
if customized ids (names) are assigned, we need to shift the default ids first to avert conflicts during reassigning ids.